### DISCUSSION TITLE
Branching of source

### DISCUSSION THREAD

**Discussion by linuxianer99** [Openvario/meta-openvario#213] [Ideas]
This discussion should be used to think about a new branching scheme for our source code ...
see #196

**Comment by linuxianer99** [comment]
As i read in all the threat before ... i would propose the following solution:

- Don't differentiate between yocto branchnes any longer ... Just have one master.
- Use tags to mark stable states in the source (so maybe head would not working sometimes, as development is ongoing)
- Releases can be done on selected tags (maybe 3.0.0 would be released, 3.0.x not ... 3.1.0 is than the next release..)

@MaxKellermann  @August2111  @kedder  Your opinion ??

**Reply by August2111** [reply]
This sounds good for me ;-)
The only thing what I want to have in addition is a branch (or a tag) which would be every time reseted to the last released and/or prereleased tag. This can avoid always changing tag name in the `ReadMe.md` - and in a pull script.

**Reply by linuxianer99** [reply]
you mean a tag like "latest" ?? Which is set always on the lastest released tag ??

**Reply by August2111** [reply]
F.e.. the name for this tracking branches could be 'release' and 'prerelease'?

**Reply by August2111** [reply]
Yes, this does I mean, Is there a difference between the tag at the really released one and the one marked as 'prerelease'?

**Reply by August2111** [reply]
Ok, that doesn't really matter - the main releases are available as images - and actually you're only interested in 'latest'. If you have additional requirements, you can always use the tag name!

**Reply by MaxKellermann** [reply]
>     * Don't differentiate between yocto branchnes any longer ... Just have one master.

Not quite - "meta-openvario" should have two branches:
- "master" is the latest bleeding edge development, with submodules pointing to whatever Yocto branch is currently being used
- "stable" is a bugfix-only stable branch, and its Yocto submodules may (or may not) point to a different (older) Yocto branch

>     * Use tags to mark stable states in the source (so maybe head would not working sometimes, as development is ongoing)

No... even "master" should not break at any time, even though we know it happens. But if it happens, it was a mistake. Such mistakes should be found and fixed before merging to master, but ... reality.

You don't use tags to mark stable states; people who want to track stable states should instead check out the "stable" branch.

>     * Releases can be done on selected tags (maybe 3.0.0 would be released, 3.0.x not ... 3.1.0 is than the next release..)

Yes. All firmware images on the website should be based on a git tag.

**Reply by linuxianer99** [reply]
@MaxKellermann  sounds resonable.

What is your suggestion to start the branches ??

1. Doing a the master branch out of the current hardknott
2. Change the repository stucture as discussed earlier
3. Create the stable branch if we have the first stable state in master ??

For the submodules:

Adding them with --remote to always get the latest version ??
Think this can also break the build and make it not reproduceable ??

**Reply by MaxKellermann** [reply]
> @MaxKellermann sounds resonable.
> 
> What is your suggestion to start the branches ??
> 
>     1. Doing a the master branch out of the current hardknott

Ideally, we should have a master branch that ties together all of the existing git history (merge commit, no filter-branch or something that modifies the commit hashes), and the top of that branch should be current hardknott.

>     2. Change the repository stucture as discussed earlier

After the "big merge", yes.

>     3. Create the stable branch if we have the first stable state in master ??

"warrior" is the current stable, isn't it? So create "stable" from that branch; as soon as master gets the first stable release, fast-forward "stable" to that release tag. That way, people can start using "stable" today, and just keep pulling, and they'll always have what this project considers the latest stable.

> For the submodules:
> 
> Adding them with --remote to always get the latest version ?? Think this can also break the build and make it not reproduceable ??

`--remote` is an `update` option, not `add`.

A submodule always refers to a specific commit id. You can update the submodule, pulling new commits from its origin, but that's always an explicit action. The `update --remote` option is one such way to do that.

**Reply by linuxianer99** [reply]
> After the "big merge", yes.

What do you mean by "big merge" ??

**Reply by MaxKellermann** [reply]
The merge commit which ties together the git history. To preserve history.

**Reply by linuxianer99** [reply]
i don't understand sorry ... 
hardknott was branched from warrior ... so if i branch from hardknott.

Which history do you want to merge ?

**Reply by MaxKellermann** [reply]
hardknott/warrior start at 6d9a70403aee896587df36c4d2f52cf70988401c (2019)
master starts at 162df53cdda96954273c65ab60fa99f352313929 (2014)
5 years of commit history got lost in the current branches.

**Reply by linuxianer99** [reply]
@MaxKellermann 
Hmmm ... we can merge this, but the histories are not related to each other ...

We started with master ... angstrom based distro ... than next ... 

With thud I started completely new using poky ... and this brachnes are current up to now ... so hardknott is based on this branch ...

i would delete master (as next has all this commits as well) and branch from current hardknott to master ..

**Reply by MaxKellermann** [reply]
OK for me - but don't delete/overwrite the old branch, rename it.

**Reply by kedder** [reply]
Having `mater` + `stable` branches + released version tags seem to be a good, well tested practice and I totally agree with switching to it. `stable` for hotfixes and minor releases, `master` for the "next" major release and bigger changes (like switching to new yocto version).

**Comment by linuxianer99** [comment]
The switch to master branch is done !!
Please do not longer use the hardknott branch !!

**Reply by August2111** [reply]
Great work, Timo and Max!
@MaxKellermann : One small question before the door is closed: I would now call the repository 'OpenVario' - and the layer - as before - 'meta-openvario' instead of 'meta-ov'. In my opinion, this corresponds more to the expected structure - and the usual naming?

**Comment by August2111** [comment]
Unfortunately I cannot find any history entries before 2019-05-05 in the new master branch, the history is still starting on this date with. the 'initial commit'.
@linuxianer99 Do you really sure that all old branches are merged?

**Reply by linuxianer99** [reply]
@August2111 

See my post here:

> Hmmm ... we can merge this, but the histories are not related to each other ...
> 
> We started with master ... angstrom based distro ... than next ...
> 
> With thud I started completely new using poky ... and this brachnes are current up to now ... so hardknott is based on this branch ...
> 
> i would delete master (as next has all this commits as well) and branch from current hardknott to master ..

The distribution before has complete different basis ... not possible to link these histories together ...

**Reply by August2111** [reply]
Sorry, I overlooked that!
